uniffi_bindgen/src/scaffolding/mod.rs

```diff
@@ -70,67 +68,55 @@ mod filters {
-                type_name
-            ),
-            _ => format!("<{} as uniffi::ViaFfi>::lower({})", type_rs(type_)?, nm),
-        })
-    }
-
-    pub fn lift_rs(nm: &dyn fmt::Display, type_: &Type) -> Result<String, askama::Error> {
-        // By explicitly naming the type here, we help the rust compiler to type-check the user-provided
-        // implementations of the functions that we're wrapping (and also to type-check our generated code).
-        // This will panic if the bindings provide an invalid value over the FFI.
-        Ok(match type_ {
-            Type::CallbackInterface(type_name) => format!(
-                "Box::new(<{}Proxy as uniffi::ViaFfi>::try_lift({}).unwrap())",
-                type_name, nm,
-            ),
-            _ => format!(
-                "<{} as uniffi::ViaFfi>::try_lift({}).unwrap()",
-                type_rs(type_)?,
-                nm
-            ),
-        })
-    }
-
-    /// Get a Rust expression for writing a value into a byte buffer.
-    pub fn write_rs(
-        nm: &dyn fmt::Display,
-        target: &dyn fmt::Display,
-        type_: &Type,
-    ) -> Result<String, askama::Error> {
-        Ok(match type_ {
-            Type::CallbackInterface(type_name) => unimplemented!(
-                "uniffi::ViaFfi::write is not supported for callback interfaces ({})",
-                type_name
-            ),
-            _ => format!(
-                "<{} as uniffi::ViaFfi>::write({}, {})",
-                type_rs(type_)?,
-                nm,
-                target
-            ),
+            // Timestamp/Duraration are handled by standard types
+            Type::Timestamp => "std::time::SystemTime".into(),
+            Type::Duration => "std::time::Duration".into(),
+            // Object is handled by Arc<T>
+            Type::Object(name) => format!("std::sync::Arc<{}>", name),
+            // Other user-defined types are handled by a unit-struct that we generate.  The
+            // FfiConverter implementation for this can be found in one of the scaffolding template code.
+            //
+            // We generate a unit-struct to sidestep Rust's orphan rules (ADR-0006).
+            //
+            // CallbackInterface is handled by special case code on both the scaffolding and
+            // bindings side.  It's not a unit-struct, but the same name generation code works.
+            Type::Enum(_) | Type::Record(_) | Type::Error(_) | Type::CallbackInterface(_) => {
+                format!("FfiConverter{}", type_.canonical_name())
+            }
+            // Wrapper types are implemented by generics that wrap the FfiConverter implementation of the
+            // inner type.
+            Type::Optional(inner) => format!("Option<{}>", ffi_converter_name(inner)?),
+            Type::Sequence(inner) => format!("Vec<{}>", ffi_converter_name(inner)?),
+            Type::Map(inner) => format!("HashMap<String, {}>", ffi_converter_name(inner)?),
```

Comment:
Aha! Coming back to this work, this is the bit that made things click for me. We're using the generated unit structs to construct an expression *in the typesystem* that knows how to convert nested structures containing imported types, without having to actually materialize any values of that type. So we might end up with `Vec<Option<FfiConverterForImportedType>>` which is not a value at runtime, but which knows how to convert a corresponding `Vec<Option<ImportedType>>` value back-and-forth across the FFI. Neat.